redis/src/aio/connection.rs

```diff
@@ -256,108 +267,205 @@ where
+    async fn command<'a>(&'a mut self, cmd: &'a Cmd) -> RedisResult<()> {
+        let mut out = Vec::new();
+
+        cmd.write_packed_command(&mut out);
+        self.framed.send(out).await?;
+
+        self.next_response().await
+    }
+
+    async fn next_response(&mut self) -> RedisResult<()> {
+        while let Some(item) = self.framed.next().await {
+            match item {
+                Ok(Ok(value)) => match Msg::from_value(&value) {
+                    Some(msg) => self.queue.push_front(msg),
+                    _ => return Ok(()),
+                },
+                _ => {}
+            }
+        }
+
+        Ok(())
+    }
+
+    /// placeholder
+    pub async fn next_message(&mut self) -> RedisResult<Msg> {
+        match self.queue.pop_back() {
+            Some(msg) => Ok(msg),
+            _ => match self.framed.next().await {
+                Some(Ok(Ok(value))) => match Msg::from_value(&value) {
+                    Some(msg) => Ok(msg),
+                    _ => Err(RedisError::from((
+                        ErrorKind::SinkDropped,
+                        "A sink corresponding to this stream has dropped",
+                    ))),
+                },
+                _ => Err(RedisError::from((
+                    ErrorKind::SinkDropped,
+                    "A sink corresponding to this stream has dropped",
+                ))),
+            },
+        }
+    }
+
+    /// placeholder
+    pub fn split(
+        self,
+    ) -> (
+        SplitPubSubSink<C>,
+        impl Stream<Item = Result<Option<Msg>, RedisError>>,
+    ) {
+        SplitPubSubSink::new(self.framed)
     }
 
     /// Subscribes to a new channel.
     pub async fn subscribe<T: ToRedisArgs>(&mut self, channel: T) -> RedisResult<()> {
-        cmd("SUBSCRIBE").arg(channel).query_async(&mut self.0).await
+        self.command(cmd("SUBSCRIBE").arg(channel)).await
     }
 
     /// Subscribes to a new channel with a pattern.
     pub async fn psubscribe<T: ToRedisArgs>(&mut self, pchannel: T) -> RedisResult<()> {
-        cmd("PSUBSCRIBE")
-            .arg(pchannel)
-            .query_async(&mut self.0)
-            .await
+        self.command(cmd("PSUBSCRIBE").arg(pchannel)).await
     }
 
     /// Unsubscribes from a channel.
     pub async fn unsubscribe<T: ToRedisArgs>(&mut self, channel: T) -> RedisResult<()> {
-        cmd("UNSUBSCRIBE")
-            .arg(channel)
-            .query_async(&mut self.0)
-            .await
+        self.command(cmd("UNSUBSCRIBE").arg(channel)).await
     }
 
     /// Unsubscribes from a channel with a pattern.
     pub async fn punsubscribe<T: ToRedisArgs>(&mut self, pchannel: T) -> RedisResult<()> {
-        cmd("PUNSUBSCRIBE")
-            .arg(pchannel)
-            .query_async(&mut self.0)
+        self.command(cmd("PUNSUBSCRIBE").arg(pchannel)).await
+    }
+}
+
+/// Represents a `PubSub` connection.
+pub struct SplitPubSubSink<C = Pin<Box<dyn AsyncStream + Send + Sync>>> {
```

Comment:
are there any tests for this? specifically, those that check that the stream is viable without the sink?

Reply:
In my opinion, if the sink drops the stream should drop.

Reply:
Can you please explain your logic?
IMO we should be consistent with other Read/Write Halves implementations, such as Unix/TcpStream. AFAIS their write halves can be dropped independently, without closing the read halves.

Reply:
I'll look at removing the limitation, I was just being consistent with the Sink requiring the Stream to work. Plus, it's an easy way to shutdown the stream which is convenient.